File: python/2020/day_19.py
import re
from functools import cache
from pathlib import Path


def main():
    text = (Path(__file__).parent / "../../input/2020/input_19.txt").read_text()
    rules, messages = text.strip().split('\n\n')
    patterns = dict(line.split(': ') for line in rules.splitlines())

    @cache
    def regex(key) -> str:
        tokens = patterns[key].split()
        for i in range(len(tokens)):
            if tokens[i].startswith('"'):
                tokens[i] = tokens[i][1:-1]
            elif tokens[i].isdigit():
                tokens[i] = f"({regex(tokens[i])})"
        return ''.join(tokens)

    print("Part 1:", sum(bool(re.fullmatch(regex('0'), line)) for line in messages.splitlines()))
    regex.cache_clear()
    patterns['8'] = "42 +"
    patterns['11'] = " | ".join(f"42 {{{n}}} 31 {{{n}}}" for n in range(1, 5))
    print("Part 2:", sum(bool(re.fullmatch(regex('0'), line)) for line in messages.splitlines()))


if __name__ == "__main__":
    main()

File: rust/2020/src/bin/day_19.rs
use regex::Regex;
use std::collections::HashMap;
use std::fs;

fn expand<'c, 'a>(
    cache: &'c mut HashMap<&'a str, String>,
    patterns: &HashMap<&str, &'a str>,
    key: &'a str,
) -> &'c str {
    if !cache.contains_key(key) {
        let expr: String = patterns[key]
            .split(' ')
            .map(|t| {
                if t.starts_with('"') {
                    // Extract letter from string
                    String::from(&t[1..2])
                } else if t.parse::<u8>().is_ok() {
                    // Recurse and expand intermediate patterns
                    format!("({})", expand(cache, patterns, t))
                } else {
                    String::from(t)
                }
            })
            .collect();
        cache.insert(key, expr);
    }
    &cache[key]
}

fn count_matches(patterns: &HashMap<&str, &str>, messages: &[&str]) -> usize {
    // Cache intermediate expansions to save time
    let mut cache = HashMap::new();
    let expr = expand(&mut cache, patterns, "0");
    let full_expr = format!("^({})$", expr);
    let re = Regex::new(&full_expr).unwrap();
    messages.iter().filter(|s| re.is_match(s)).count()
}

fn main() {
    let input = fs::read_to_string("../input/2020/input_19.txt").unwrap();
    let blocks: Vec<_> = input.split("\n\n").collect();
    let mut patterns: HashMap<_, _> = blocks[0]
        .lines()
        .map(|s| {
            let mut kv = s.split(": ");
            (kv.next().unwrap(), kv.next().unwrap())
        })
        .collect();
    let messages: Vec<_> = blocks[1].lines().collect();
    println!("Part 1: {}", count_matches(&patterns, &messages));
    // Make pattern replacements for part 2
    patterns.insert("8", "42 +");
    // Use a regular version of non-regular rule 11
    let parts: Vec<_> = (1..5)
        .map(|n| format!("42 {{{}}} 31 {{{}}}", n, n))
        .collect();
    let nonreg = parts.join(" | ");
    patterns.insert("11", &nonreg);
    println!("Part 2: {}", count_matches(&patterns, &messages));
}
